src/boot/efi/xbootldr.c

```diff
@@ -16,31 +21,163 @@ static EFI_DEVICE_PATH *path_parent(EFI_DEVICE_PATH *path, EFI_DEVICE_PATH *node
+                EFI_GUID *ret_part_uuid) {
+
+        _cleanup_freepool_ EFI_PARTITION_ENTRY *entries = NULL;
+        union GptHeaderBuffer gpt;
+        EFI_STATUS err;
+        UINT32 crc32;
+        UINTN size;
+
+        assert(block_io);
+        assert(ret_part_number);
+        assert(ret_part_start);
+        assert(ret_part_size);
+        assert(ret_part_uuid);
+
+        /* Read the GPT header */
+        err = uefi_call_wrapper(
+                        block_io->ReadBlocks, 5,
+                        block_io,
+                        block_io->Media->MediaId,
+                        lba,
+                        sizeof(gpt), &gpt);
+        if (EFI_ERROR(err))
+                return err;
+
+        if (!verify_gpt(&gpt, lba))
+                return EFI_DEVICE_ERROR;
+
+        /* Now load the GPT entry table */
+        size = ALIGN_TO((UINTN) gpt.gpt_header.SizeOfPartitionEntry * (UINTN) gpt.gpt_header.NumberOfPartitionEntries, 512);
+        entries = AllocatePool(size);
+        if (!entries)
+                return EFI_OUT_OF_RESOURCES;
+
+        err = uefi_call_wrapper(
+                        block_io->ReadBlocks, 5,
+                        block_io,
+                        block_io->Media->MediaId,
+                        gpt.gpt_header.PartitionEntryLBA,
+                        size, entries);
+        if (EFI_ERROR(err))
+                return err;
+
+        /* Calculate CRC of entries array, too */
+        err = BS->CalculateCrc32(entries, size, &crc32);
+        if (EFI_ERROR(err) || crc32 != gpt.gpt_header.PartitionEntryArrayCRC32)
+                return err;
+
+        /* Now we can finally look for xbootloader partitions. */
+        for (UINTN i = 0; i < gpt.gpt_header.NumberOfPartitionEntries; i++) {
+                EFI_PARTITION_ENTRY *entry;
+                EFI_LBA start, end;
+
+                entry = (EFI_PARTITION_ENTRY*) ((UINT8*) entries + gpt.gpt_header.SizeOfPartitionEntry * i);
+
+                if (CompareMem(&entry->PartitionTypeGUID, XBOOTLDR_GUID, sizeof(entry->PartitionTypeGUID)) != 0)
+                        continue;
+
+                /* Let's use memcpy(), in case the structs are not aligned (they really should be though) */
+                CopyMem(&start, &entry->StartingLBA, sizeof(start));
+                CopyMem(&end, &entry->EndingLBA, sizeof(end));
+
+                if (end < start) /* Bogus? */
+                        continue;
+
+                *ret_part_number = i + 1;
+                *ret_part_start = start;
+                *ret_part_size = end - start + 1;
+                CopyMem(ret_part_uuid, &entry->UniquePartitionGUID, sizeof(*ret_part_uuid));
+
+                return EFI_SUCCESS;
+        }
+
+        /* This GPT was fully valid, but we didn't find what we are looking for. This
+         * means there's no reason to check the second copy of the GPT header */
+        return EFI_NOT_FOUND;
+}
+
+static EFI_DEVICE_PATH *find_device(
```

Comment:
given this might result in errors, i'd really use a prototype that returns EFI_STATUS, and returns the EFI_DEVICE_PATH object as yet another return param